File: base/src/main/java/de/whitefrog/frogr/patch/Patch.java
package de.whitefrog.frogr.patch;

import com.github.zafarkhaja.semver.Version;
import de.whitefrog.frogr.Service;

/**
 * Database patch for bulk operations and field changes.
 */
public abstract class Patch implements Comparable<Patch> {
  private Service service;
  private final Version version;
  private final int priority;

  public Patch(Service service) {
    this.service = service;
    this.version = Version.valueOf(getClass().getAnnotation(de.whitefrog.frogr.patch.Version.class).value());
    this.priority = getClass().getAnnotation(de.whitefrog.frogr.patch.Version.class).proority();
  }

  public int getPriority() {
    return priority;
  }

  public Service getService() {
    return service;
  }

  public abstract void update();

  public Version getVersion() {
    return version;
  }

  @Override
  public int compareTo(Patch o) {
    if(getVersion().greaterThan(o.getVersion()) || getPriority() < o.getPriority()) return 1;
    else if(getVersion().equals(o.getVersion()) && getPriority() == o.getPriority()) return 0;
    else return -1;
  }
}

File: base/src/main/java/de/whitefrog/frogr/patch/Patch.kt
package de.whitefrog.frogr.patch

import com.github.zafarkhaja.semver.Version
import de.whitefrog.frogr.Service

/**
 * Database patch for bulk operations and field changes.
 */
abstract class Patch(val service: Service) : Comparable<Patch> {
  val version: Version = Version.valueOf(javaClass.getAnnotation(de.whitefrog.frogr.patch.Version::class.java).value)
  val priority: Int = javaClass.getAnnotation(de.whitefrog.frogr.patch.Version::class.java).priority

  abstract fun update()

  override fun compareTo(o: Patch): Int {
    return if (version.greaterThan(o.version) || priority < o.priority)
      1
    else if (version == o.version && priority == o.priority)
      0
    else
      -1
  }
}
